packages/nextcloud/lib/src/interceptors/http_interceptor.dart

```diff
@@ -0,0 +1,25 @@
+import 'dart:async';
+
+import 'package:http/http.dart' as http;
+
+/// Interceptor that can manipulate http requests and responses.
+///
+/// The interface is inspired by the one `InterceptorContract` from
+/// the [http_interceptor](https://pub.dev/packages/http_interceptor)
+/// package for easy adaption between other existing interceptors.
+abstract class HttpInterceptor<S extends http.BaseRequest, T extends http.BaseResponse> {
+  /// Whether this interceptor should intercept requests.
+  bool shouldInterceptRequest() => true;
+
+  /// Intercepts the given [request].
+  ///
+  /// Provided requests are not finalized yet. It is an error for an interceptor
+  /// to request itself.
+  FutureOr<S> interceptRequest({required S request});
+
+  /// Whether this interceptor should intercept response.
+  bool shouldInterceptResponse() => true;
+
+  /// Intercepts the given [response].
+  FutureOr<T> interceptResponse({required T response});
```

Comment:
But this can be used for caching, no?

Reply:
Ah no I was wrong...

Reply:
Not really.
If a cache interceptor has a stored value there is currently no way to bypass the actual http request being made and return the cached one.

Reply:
Yeah but we can easly add two more methods to intercept the actual request and return a response. I'll do that later once it is necessary.